### PR TITLE
[move][ide][move-compiler] "typing symbolicator" into a typing analysis visito

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

Converts the move analyzer "typing symbolicator" into a typing analysis visitor that uses a (much-extended) typing visitor to extract information.

Note that the extensions to the typing visitor are opt-in via constants defined on the trait, and Rust's optimizer appears to completely elide that code when the constants are false.

## Test plan 

All tests still work as expected (though I did not test the actual IDE usage -- maybe @awelc  can)

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs [hunk 1]
```diff
@@ -35,10 +37,36 @@ pub trait TypingVisitorConstructor {
...
+    const VISIT_TYPES: bool = false;
+
+    /// Indicates if lvalues should be visited during the traversal of sequence forms.
+    const VISIT_LVALUES: bool = false;
+
+    /// Indicates if use_funs should be visited during the traversal.
+    const VISIT_USE_FUNS: bool = false;
+
+    /// By default, the visitor will visit all all expressions in all functions in all modules. A
+    /// custom version should of this function should be created if different type of analysis is
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/typing/visitor.rs:60]:
```suggestion
    /// custom version of this function should be created if different type of analysis is
```

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs [hunk 2]
```diff
@@ -163,24 +239,156 @@ pub trait TypingVisitorContext {
...
-    fn visit_seq(&mut self, (_, seq): &mut T::Sequence) {
+    // -- TYPES --
+
+    fn visit_type_custom(&mut self, _ty: &mut N::Type) -> bool {
+        false
+    }
+
+    /// Visit a type, including recursively. Note that this may be called manually even if
+    /// `VISIT_TYPES` is set to `false`.
+    #[growing_stack]
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/typing/visitor.rs:264]: When do we need to include the `#[growing_stack]` annotations? Is it related to recursion (though it's not specified for `visit_lvalue` so probably not...)?

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs [hunk 3]
```diff
@@ -0,0 +1,942 @@
...
+            T::FunctionBody_::Macro | T::FunctionBody_::Native => (),
+        }
+        // process return types
+        self.visit_type(&mut fdef.signature.return_type);
+
+        // clear type params from the scope
+        self.type_params.clear();
+    }
+
+    fn visit_lvalue_list(&mut self, kind: &LValueKind, lvalues: &mut T::LValueList) {
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:665]: Isn't it handled by the default visitor implementation?

**Review by awelc — APPROVED** [review]
Other than some minor comments, it looks good. It may not make the analyzer-side code size that much smaller, but it does make it look a lot cleaner! 

In terms of testing, this type of change would very likely trigger a unit test failure (which it does does not) but I also tested this manually on Sui framework code to see if nothing funky is going on.

Stamping but should likely wait for @tnowacki's approval

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/typing/visitor.rs:264]: We generally do when recurring. it should be on `visit_lvalue`.

File: external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs (hunk 3 above)
Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs:665]: Yes. I had my IDE inline all the functions and deleted all the ones I didn't want to change the behavior of, but missed this one.

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs [hunk 4]
```diff
@@ -163,24 +241,149 @@ pub trait TypingVisitorContext {
...
             SI::Seq(e) => self.visit_exp(e),
+            SI::Declare(lvalues) if Self::VISIT_LVALUES => {
+                self.visit_lvalue_list(&LValueKind::Bind, lvalues);
+            }
             SI::Declare(_) => (),
-            SI::Bind(_, _, e) => self.visit_exp(e),
+            SI::Bind(lvalues, ty_ann, e) => {
+                // visit the RHS first to better match control flow
+                self.visit_exp(e);
+                if Self::VISIT_LVALUES {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/typing/visitor.rs:328]: Should we put these sort of gates inside of the respective functions, e.g. `self.visit_lvalue_list` checks `if !Self::VISIT_LVALUES { return }`

Will Rust still be able to compile these calls away?

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs [hunk 5]
```diff
@@ -195,11 +398,31 @@ pub trait TypingVisitorContext {
         if self.visit_exp_custom(exp) {
             return;
         }
+        if Self::VISIT_TYPES {
+            self.visit_type(&mut exp.ty);
+        }
         let sp!(_, uexp) = &mut exp.exp;
         match uexp {
             E::ModuleCall(c) => self.visit_exp(&mut c.arguments),
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/typing/visitor.rs:406]: I think we need to visit the type arguments here?

**Review by tnowacki — APPROVED** [review]
Approved, thanks! 
We might want to do this for the other visitors at some point :) 

Two concerns:
1) it is a bit unfortunate we have to gate the calls, rather than gating inside of the call itself. I think it is worth seeing if Rust would compile these calls away if the gating was in the function--rather than at the call site. 
2) I'm worried about the structure of the visitors today. As we add more and more of them, we might take a bit of a perf hit, since it has to walk the whole AST n-times, rather than visiting each element n-times. Maybe it won't make a big difference, but it might? In order to do this change, the recursion/traversal would have to be separated from the custom-visit functions. Not sure if I'm making sense, but we can talk about it later

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs (hunk 4 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/typing/visitor.rs:328]: Rust might erase the calls if they are gated and inlining catches / erases them. That doesn't happen in debug mode, but may in release mode. My other motivation for gating them externally is that someone can still call, e.g., `visit_type` in a non-`VISIT_TYPES` visitor if they want to in just a few locations, allowing for a sliver more flexibility.

File: external-crates/move/crates/move-compiler/src/typing/visitor.rs (hunk 5 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/typing/visitor.rs]: And the parameter types! Thanks for catching this.

**Pull request merged**